core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,430 @@
+
+              currOffset = batch.nextOffset
+            }
+          }
+
+          loadedTransactions.foreach {
+            case (transactionalId, txnMetadata) =>
+              val currentTxnMetadata = addTransaction(transactionalId, txnMetadata)
+              if (!txnMetadata.equals(currentTxnMetadata)) {
+                // treat this as a fatal failure as this should never happen
+                fatal(s"Attempt to load $transactionalId's metadata $txnMetadata failed " +
+                  s"because there is already a different cached transaction metadata $currentTxnMetadata.")
+
+                throw new KafkaException("Loading transaction topic partition failed.")
+              }
+          }
+
+          removedTransactionalIds.foreach { transactionalId =>
+            if (transactionMetadataCache.contains(transactionalId)) {
+              // the cache already contains a transaction which should be removed,
+              // treat this as a fatal failure as this should never happen
+              fatal(s"Unexpected to see $transactionalId's metadata while " +
+                s"loading partition $topicPartition since its latest state is a tombstone")
+
+              throw new KafkaException("Loading transaction topic partition failed.")
+            }
+          }
+
+          info(s"Finished loading ${loadedTransactions.size} transaction metadata from $topicPartition in ${time.milliseconds() - startMs} milliseconds")
+        }
+    }
+  }
+
+  /**
+   * When this broker becomes a leader for a transaction log partition, load this partition and
+   * populate the transaction metadata cache with the transactional ids.
+   */
+  def loadTransactionsForPartition(partition: Int, coordinatorEpoch: Int) {
+    validateTransactionTopicPartitionCountIsStable()
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    inLock(stateLock) {
+      ownedPartitions.put(partition, coordinatorEpoch)
+      loadingPartitions.add(partition)
+    }
+
+    def loadTransactions() {
+      info(s"Loading transaction metadata from $topicPartition")
+      try {
+        loadTransactionMetadata(topicPartition)
+      } catch {
+        case t: Throwable => error(s"Error loading transactions from transaction log $topicPartition", t)
+      } finally {
+        inLock(stateLock) {
+          loadingPartitions.remove(partition)
+        }
+      }
+    }
+
+    scheduler.schedule(topicPartition.toString, loadTransactions)
+  }
+
+  /**
+   * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
+   * that belong to that partition.
+   */
+  def removeTransactionsForPartition(partition: Int) {
+    validateTransactionTopicPartitionCountIsStable()
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    inLock(stateLock) {
+      ownedPartitions.remove(partition)
+    }
+
+    def removeTransactions() {
```

Comment:
I guess it is because `loadTransactions` is run in the scheduler. So they won't interleave. @guozhangwang ?